tianshou/utils/net/common.py

```diff
@@ -595,3 +598,13 @@ def forward(self, obs: np.ndarray | torch.Tensor, *args, **kwargs) -> Any:
         return new_net_class
 
     return decorator_fn, new_state_shape
+
+
+class BaseActor(nn.Module, ABC):
```

Comment:
Maybe instead `ActorProtocol` with fields `preprocess` and `output_dim`? I don't feel like the new getter methods that return public attributes are adding value

Reply:
You could inherit from the protocol, to make it explicit. My point was not to introduce duck typing but rather to not bloat the public interface by adding rather redundant methods. It can't also be achieved in a different way:

1. Make the public field private. You could then also use abstract properties in the base class instead of getters
2. Let the abstract class have an init that creates these attributes

I would lean towards the second solution as it needs less refactoring, and I remember already doing something like this with the critics

Reply:
Inheriting from a protocol does not really work, as it does not enforce that the protocol is actually implemented - and no one (i.e. no dev tool I am aware of) will complain if it's not or offer you an auto-completion that will help you get it done.

I don't see how changing the base class to have state, necessitating a super call in init, improves anything. Without it, subclasses are free to choose any internal representation and simply implement the two methods. Importantly, using public properties rather than methods is not generally viewed as a preferable interface; this may be your personal preference, but it does not align with general conventions in OOP. From the point of view of the abstraction `BaseActor`, there is a clearly defined interface based on methods. The fact that the interface could also take another form when looking at the subclasses (all of which happen to use public attributes) is not relevant nor a problem.

In general, the reason why methods are preferred is that internal representations often change, and many programmers carelessly use public attributes and later break the (supposedly public) interface by changing the attributes. A method-based interface can be more stable (and is perceived as reliable/permanent); For this reason, the convention to never use any public properties except in dataclass-style representations is frequently adhered to. (Python's property annotation can help provide stable interfaces when using properties, of course, but the fact remains that a public property just doesn't really mean much to most Python programmers.)

Reply:
Ok, it's fine for me leaving it as is. But does anything speak against making the corresponding attributes private in the implementations? Could be done later as well, course

Reply:
I wouldn't mind making them private, but it constitutes a breaking change.

Reply:
Resolving, since it's out of scope for this PR and you don't have much time. At some point in the future we can think about improving the structure of the public interfaces

Reply:
Yeah, output_dim is really hacky in the current implementation. I totally agree we should make it as a standard interface.